sys/kern/sleepq.c

```diff
@@ -216,18 +216,33 @@ static void sq_leave(thread_t *td, sleepq_chain_t *sc, sleepq_t *sq) {
 static int _sleepq_wait(void *wchan, const void *waitpt, sleep_t sleep) {
   thread_t *td = thread_self();
   int status = 0;
+#define EARLY_INTR(td, sleep)                                                  \
+  (((td)->td_flags & TDF_NEEDSIGCHK) != 0 && sleep == SLP_INTR)
+
+  /* If there are pending signals, interrupt the sleep immediately. */
+  WITH_SPIN_LOCK (&td->td_spin) {
+    if (EARLY_INTR(td, sleep))
+      return EINTR;
+  }
 
   if (waitpt == NULL)
     waitpt = __caller(0);
 
   sq_enter(td, wchan, waitpt, sleep);
 
   /* The code can be interrupted in here.
-   * A race is avoided by clever use of TDF_SLEEPY flag. */
+   * A race is avoided by clever use of TDF_SLEEPY flag.
+   * We can also get a signal in here -- interrupt early if we got one.
+   * The first signal check is an optimization that saves us the call
+   * to sq_enter. */
 
   WITH_SPIN_LOCK (&td->td_spin) {
     if (td->td_flags & TDF_SLEEPY) {
       td->td_flags &= ~TDF_SLEEPY;
+      if (EARLY_INTR(td, sleep)) {
+        td->td_flags &= ~(TDF_SLPINTR | TDF_SLPTIMED);
+        return EINTR;
```

Comment:
At some point I'll remove all references to obscure `WITH_*` macros.
I'd much prefer to have a single exit point from the critical section.